src/disjoint_mut.rs

```diff
@@ -0,0 +1,417 @@
+                    panic!("{range:?} on thread {thread:?} overlaps with an existing mutable range: {r:#?}");
+                }
+            }
+            self.ranges
+                .immutable
+                .lock()
+                .unwrap()
+                .push(DisjointMutRange::new(range.to_bounds()));
+        }
+
+        fn remove_range(&self, range: (Bound<usize>, Bound<usize>), mutable: bool) {
+            let mut ranges = if mutable {
+                self.ranges.mutable.lock().unwrap()
+            } else {
+                self.ranges.immutable.lock().unwrap()
+            };
+            let idx = ranges
+                .iter()
+                .position(|r| r == &range)
+                .expect("Expected range {range:?} to be in the active ranges");
+            ranges.remove(idx);
+        }
+    }
+
+    pub trait DisjointMutIndex: Debug {
+        fn to_bounds(&self) -> (Bound<usize>, Bound<usize>);
+        fn contains<R: RangeBounds<usize>>(&self, other: &R) -> bool;
+    }
+
+    impl DisjointMutIndex for usize {
+        fn to_bounds(&self) -> (Bound<usize>, Bound<usize>) {
+            (Bound::Included(*self), Bound::Excluded(self + 1))
+        }
+
+        fn contains<R: RangeBounds<usize>>(&self, other: &R) -> bool {
+            other.contains(self)
+        }
+    }
+
+    macro_rules! impl_disjoint_mut_index {
+        ($range_type:ty) => {
+            impl DisjointMutIndex for $range_type {
+                fn to_bounds(&self) -> (Bound<usize>, Bound<usize>) {
+                    (self.start_bound().cloned(), self.end_bound().cloned())
+                }
+
+                fn contains<R: RangeBounds<usize>>(&self, other: &R) -> bool {
+                    match (other.start_bound(), self.end_bound()) {
+                        (Bound::Included(i), Bound::Included(j)) => {
+                            if i > j {
+                                return false;
+                            }
+                        }
+                        (Bound::Included(i), Bound::Excluded(j))
+                        | (Bound::Excluded(i), Bound::Excluded(j))
+                        | (Bound::Excluded(i), Bound::Included(j)) => {
+                            if i >= j {
+                                return false;
+                            }
+                        }
+                        (Bound::Unbounded, _) => {}
+                        (_, Bound::Unbounded) => {}
+                    }
+                    match (other.end_bound(), self.start_bound()) {
+                        (Bound::Included(i), Bound::Included(j)) => i >= j,
+                        (Bound::Included(i), Bound::Excluded(j))
+                        | (Bound::Excluded(i), Bound::Included(j)) => i > j,
+                        (Bound::Excluded(i), Bound::Excluded(j)) => i - 1 > *j,
+                        (Bound::Unbounded, _) | (_, Bound::Unbounded) => true,
+                    }
+                }
```

Comment:
I was having a hard time understanding this logic enough to confirm that it's accurate, so I ended up writing a unit test to check various edge cases:

```rust
#[test]
fn test_range_overlap() {
    // Range overlap.
    assert!((5..7).overlaps(&(4..10)));
    assert!((4..10).overlaps(&(5..7)));
    
    // RangeFrom overlap.
    assert!((5..).overlaps(&(4..10)));
    assert!((4..10).overlaps(&(5..)));

    // RangeTo overlap.
    assert!((..7).overlaps(&(4..10)));
    assert!((4..10).overlaps(&(..7)));

    // RangeInclusive overlap.
    assert!((5..=7).overlaps(&(7..10)));
    assert!((7..10).overlaps(&(5..=7)));

    // RangeToInclusive overlap.
    assert!((..=7).overlaps(&(7..10)));
    assert!((7..10).overlaps(&(..=7)));

    // Range no overlap.
    assert!(!(5..7).overlaps(&(10..20)));
    assert!(!(10..20).overlaps(&(5..7)));

    // RangeFrom no overlap.
    assert!(!(15..).overlaps(&(4..10)));
    assert!(!(4..10).overlaps(&(15..)));

    // RangeTo no overlap.
    assert!(!(..7).overlaps(&(10..20)));
    assert!(!(10..20).overlaps(&(..7)));

    // RangeInclusive no overlap.
    assert!(!(5..=7).overlaps(&(8..10)));
    assert!(!(8..10).overlaps(&(5..=7)));

    // RangeToInclusive no overlap.
    assert!(!(..=7).overlaps(&(8..10)));
    assert!(!(8..10).overlaps(&(..=7)));
}
```

I think it'd be nice to include a unit test like this for potentially tricky logic, though we're not running Rust tests in CI so it'd only be helpful for manually verifying changes. I also included tests for `RangeInclusive` and `RangeToInclusive` in case you decide to add support for those. Up to you if you want to add it or not.